### PR TITLE
DSS Infra: Terraforming

### PR DESCRIPTION
**Pull request by amarjandu**
Allows faster auth0 deployments for the datastore

### CODE REVIEW

File: environment [hunk 1]
```diff
@@ -18,14 +18,10 @@ DSS_SECRETS_STORE="${DSS_PLATFORM}/dss"
...
 DSS_INFRA_TAG_SERVICE="dss"
-DSS_INFRA_TAG_OWNER="team-redwood-group@ucsc.edui"
-
-#not sure if these are needed
-OIDC_AUDIENCE=https://dss.dev.ucsc-cgp-redwood.org
-AUTH_URL=https://auth.ucsc.ucsc-cgp-redwood.org
-OPENID_PROVIDER=https://dev-4lyab62k.auth0.com/
-OIDC_EMAIL_CLAIM="${OIDC_AUDIENCE}/email"
-OIDC_GROUP_CLAIM="${OIDC_AUDIENCE}/group"
+DSS_INFRA_TAG_OWNER="team-redwood-group@ucsc.edu"
```
Comment (chmreid) [inline environment]:
```suggestion
DSS_INFRA_TAG_OWNER="team-redwood@ucsc.edu"
```

Comment (chmreid) [reply environment]: Per https://github.com/DataBiosphere/data-store/issues/46

File: infra/build_deploy_config.py [hunk 2]
```diff
@@ -77,6 +77,7 @@ def get_auth_tenant_config():
     "DSS_INFRA_TAG_SERVICE",
     "DSS_SECRETS_STORE",
     "GCP_DEFAULT_REGION",
+    "OIDC_AUDIANCE"
```
Comment (chmreid) [inline infra/build_deploy_config.py]:
```suggestion
    "OIDC_AUDIENCE"
```

File: infra/auth/main.tf [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+resource "auth0_rule" "custom_claim" {
+  name = "jwt_claim_from_domain"
+  // this script might need to be migrated out of this and into the build script....
+  script = <<EOF
+function (user, context, callback) {
+    var privileged_domains = ["ucsc.edu","platform-hca.iam.gserviceaccount.com"]; //authorized domains
+    var domain = user.email.split('@').pop();
+    const namespace = "${var.OIDC_AUDIANCE}";
```
Comment (chmreid) [inline infra/auth/main.tf]:
```suggestion
    const namespace = "${var.OIDC_AUDIENCE}";
```

File: environment [hunk 4]
```diff
@@ -18,14 +18,10 @@ DSS_SECRETS_STORE="${DSS_PLATFORM}/dss"
...
+
+# dont use interpolation on the stage here, its going to change the claims that are set in the auth0_rule
+OIDC_AUDIANCE=https://dss.dev.ucsc-cgp-redwood.org
```
Comment (chmreid) [inline environment]:
```suggestion
OIDC_AUDIENCE=https://dss.dev.ucsc-cgp-redwood.org
```

**Review by chmreid — CHANGES_REQUESTED** [review]
Please see suggested changes

File: environment [hunk 5]
```diff
@@ -18,14 +18,10 @@ DSS_SECRETS_STORE="${DSS_PLATFORM}/dss"
...
-
-#not sure if these are needed
-OIDC_AUDIENCE=https://dss.dev.ucsc-cgp-redwood.org
-AUTH_URL=https://auth.ucsc.ucsc-cgp-redwood.org
-OPENID_PROVIDER=https://dev-4lyab62k.auth0.com/
-OIDC_EMAIL_CLAIM="${OIDC_AUDIENCE}/email"
-OIDC_GROUP_CLAIM="${OIDC_AUDIENCE}/group"
+DSS_INFRA_TAG_OWNER="team-redwood-group@ucsc.edu"
+
+# dont use interpolation on the stage here, its going to change the claims that are set in the auth0_rule
```
Comment (chmreid) [inline environment:23]: 👍  tnx for leaving helpful comments

File: infra/auth/main.tf [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+resource "auth0_client" "dss_auth" {
+  name = "data-store-auth"
+  description = "simple auth system for the data-store; managed by Terraform"
+  app_type = "regular_web"
+  is_first_party = true
+  is_token_endpoint_ip_header_trusted = true
+  token_endpoint_auth_method = "client_secret_post"
+  oidc_conformant = true
+  callbacks = [ "https://example.com/callback", "http://localhost:8080" ]
```
Comment (chmreid) [inline infra/auth/main.tf]: Should these `example.com` values be updated?

File: infra/auth/main.tf [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+resource "auth0_connection" "google_connector" {
+  # tie this into google provider for IdP at a later time
+  # need to sort out domain ownership....
+  name = "google"
+  strategy = "google-oauth2"
+  enabled_clients = ["${auth0_client.dss_auth.client_id}"]
+}
+
+resource "auth0_rule" "custom_claim" {
```
Comment (chmreid) [inline infra/auth/main.tf:55]: This is awesome

File: requirements.txt [hunk 8]
```diff
@@ -1,3 +1,13 @@
+<<<<<<< HEAD
```
Comment (chmreid) [inline requirements.txt]: oops, looks like you committed some merge conflict stuff

**Review by chmreid — CHANGES_REQUESTED** [review]
Looking good so far, that script to set the custom auth0 claim is slick.

A few things to address in comments, then I'll give this one more review

File: infra/auth/main.tf (hunk 6 above)
Comment (amarjandu) [reply infra/auth/main.tf]: done!

File: environment (hunk 5 above)
Comment (amarjandu) [reply environment:23]: its for us in the future tbh.
in the future we gonna look back and say, `thx amar`

File: requirements.txt (hunk 8 above)
Comment (amarjandu) [reply requirements.txt]: removed!

File: infra/auth/main.tf (hunk 7 above)
Comment (amarjandu) [reply infra/auth/main.tf:55]: 😎

File: environment (hunk 5 above)
Comment (charlesreid1) [reply environment]: May want to mention OIDC_AUDIENCE variable should have a trailing slash

File: environment [hunk 9]
```diff
@@ -18,15 +18,13 @@ DSS_SECRETS_STORE="${DSS_PLATFORM}/dss"
...
 DSS_INFRA_TAG_PROJECT="${DSS_PLATFORM}-dss"
 DSS_INFRA_TAG_SERVICE="dss"
-DSS_INFRA_TAG_OWNER="team-redwood-group@ucsc.edui"
+DSS_INFRA_TAG_OWNER="team-redwood-group@ucsc.edu"
 
-#not sure if these are needed
-OIDC_AUDIENCE=https://dss.dev.ucsc-cgp-redwood.org
-AUTH_URL=https://auth.ucsc.ucsc-cgp-redwood.org
+# dont use interpolation on the stage here, its going to change the claims that are set in the auth0_rule
+# Warning back slashes are sensitive here
```
Comment (charlesreid1) [inline environment:24]: May want to be a little more clear what you mean here

Would be helpful to mention OIDC_AUDIENCE needs a trailing slash

**Review by charlesreid1 — APPROVED** [review]

File: environment (hunk 9 above)
Comment (chmreid) [reply environment:24]: I agree wholeheartedly with this very stable genius who is making random suggestions for us in our repo

**Review by chmreid — APPROVED** [review]

**Pull request merged**